java/feathub-udf/flink-udf/src/main/java/com/alibaba/feathub/flink/udf/processfunction/SlidingWindowKeyedProcessFunction.java

```diff
@@ -333,23 +391,60 @@ public static SlidingWindowState create(
 
             return new SlidingWindowState(
                     aggregationFieldsDescriptor,
-                    timestampToRow,
+                    timestampToRows,
                     listState,
                     outFieldNameToLeftTimestampIdx,
                     maxRegisteredTimer,
+                    maxTriggeredTimer,
                     lastOutputRow,
                     accumulatorStates);
         }
 
         /**
-         * Add the row to the state.
+         * Add the row to the state and adjust left idx accordingly.
          *
          * @param timestamp The row time of the row.
-         * @param row The row to be added
+         * @param row The row to be added.
+         * @param affectedAggFieldIndices The indices of the agg fields that would add or have added
+         *     this row to their accumulator. If this value is null, it means that all agg fields
+         *     add this row.
          */
-        public void addRow(long timestamp, Row row) throws Exception {
-            timestampToRow.put(timestamp, row);
-            timestampList.add(timestamp);
+        public void addRow(long timestamp, Row row, Set<Integer> affectedAggFieldIndices)
+                throws Exception {
+            if (timestampToRows.contains(timestamp)) {
+                List<Row> rows = timestampToRows.get(timestamp);
+                rows.add(row);
+                timestampToRows.put(timestamp, rows);
+            } else {
+                timestampToRows.put(timestamp, new ArrayList<>(Collections.singletonList(row)));
+
+                List<Long> timestamps = getTimestampList();
+
+                if (timestamps.isEmpty() || timestamps.get(timestamps.size() - 1) < timestamp) {
+                    timestampList.add(timestamp);
+                } else {
+                    for (int i = 0; i < timestamps.size(); i++) {
+                        if (timestamps.get(i) < timestamp) {
+                            continue;
+                        }
+                        timestamps.add(i, timestamp);
+                        break;
+                    }
+                    timestampList.update(timestamps);
+                }
+
+                List<Integer> leftIdxList = getLeftTimestampIdx();
+                for (AggregationFieldsDescriptor.AggregationFieldDescriptor aggFieldDescriptor :
+                        aggregationFieldsDescriptor.getAggFieldDescriptors()) {
+                    int aggFieldIdx =
+                            aggregationFieldsDescriptor.getAggFieldIdx(aggFieldDescriptor);
+                    if (affectedAggFieldIndices != null
+                            && !affectedAggFieldIndices.contains(aggFieldIdx)) {
+                        leftIdxList.set(aggFieldIdx, leftIdxList.get(aggFieldIdx) + 1);
```

Comment:
- After inserting the timestamp and row into states, we need to update left idx list so that left indices can correctly reflect the rows to retract.
- For each aggregation field, we need to compare `row time` and `maxTriggeredTimer - windowSizeMs` to judge whether `leftIdx` should increment by 1 or not. We cannot directly compare `leftIdx` and the row's insertion index(`insertIdx`), because when `leftIdx == insertIdx`, both cases are possible.
- The comparison of `row time` and `maxTriggeredTimer - windowSizeMs` has been performed in `processElement`, so I choose to record that result in `affectedAggFieldIndices` variable and reuse the result here.